endorsedLibraries/kotlinx.cli/src/main/kotlin/kotlinx/cli/Arguments.kt

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+package kotlinx.cli
+
+import kotlin.reflect.KProperty
+
+/**
+ * Interface for CLI.
+ */
+interface CLIEntity<T : Any, TResult> {
+    /**
+     * Argument value.
+     */
+    val value: ArgumentValueDelegate<TResult>
+    /**
+     * Origin of argument value.
+     */
+    val valueOrigin: ArgParser.ValueOrigin
+        get() = (value as ParsingValue<*, *>).valueOrigin
+
+    operator fun provideDelegate(thisRef: Any?, prop: KProperty<*>): ArgumentValueDelegate<TResult>
+}
+
+/**
+ * Command line entity.
+ *
+ * @property owner parser which owns current entity.
+ */
+internal abstract class CLIEntityImpl<T : Any, TResult>(val owner: ArgParser): CLIEntity<T, TResult> {
+    /**
+     * Unique id.
+     */
+    var id: Int? = null
+    /**
+     * CLI element value.
+     */
+    internal lateinit var cliElement: ArgumentValueDelegate<TResult>
+    /**
+     * Wrapper  for element - read only property.
+     * Needed to close set of variable [cliElement].
+     */
+    override val value: ArgumentValueDelegate<TResult>
+        get() = cliElement
+}
+
+/**
+ * Argument instance interface.
+ */
+interface Argument<T : Any, TResult>: CLIEntity<T, TResult>
```

Comment:
Separate `Argument` and `Option` interfaces to save opportunity provide different modifiers now and in future.